KHR audio framework design proposal (#2421)
KhronosGroup/glTF: extensions/2.0/Khronos/KHR_audio_graph/README.md

<!-----

Conversion time: 1.967 seconds.

Using this Markdown file:

1. Paste this output into your source file.
2. See the notes and action items below regarding this conversion run.
3. Check the rendered output (headings, lists, code blocks, tables) for proper

>    formatting and use a linkchecker before you publish this page.

Reviewer: Is this leftover from something? It should be removed.

Conversion notes:

* Docs to Markdown version 1.0β36
* Fri Jun 14 2024 17:25:52 GMT-0700 (PDT)
* Source doc: KHR Audio Graph Design
* Tables are currently converted to HTML tables.
----->

## KHR Audio Graph Design

> ## **Contributors**

Reviewer: Headings shouldn't have any additional formatting symbols.
```suggestion
## Contributors
```

> * Chintan Shah, Meta
> * Alexey Medvedev, Meta

Reviewer: The Khronos preferred Markdown style is to use `-` for unordered lists:
```suggestion
- Chintan Shah, Meta
- Alexey Medvedev, Meta
```

## Context

> During the recent Khronos 3D formats working group meeting held on 5/29, we reviewed the [proposal to define the KHR audio glTF specification using an audio graph framework](https://docs.google.com/presentation/d/1IrrQaE-jHyzOtFRabjtLAzeP5UOirFOEj8FRADAceqk/edit?usp=sharing). The purpose of this document is to delve deeper into that proposal, offering a comprehensive design of the KHR audio graph. This includes a detailed description of each node object within the graph along with functionality, the specific properties associated with it, and how it interacts with other nodes in the graph. The document is structured to facilitate clear understanding and to solicit feedback on the proposed design. Based on the feedback we will update and finalize the design before it is formally schematized into KHR core audio spec, extensions, animations, and interactivity.

Reviewer: Please include the year and use ISO 8601 to make this unambiguous to future readers.

```suggestion
During the recent Khronos 3D formats working group meeting held on 2024-05-29, we reviewed the [proposal to define the KHR audio glTF specification using an audio graph framework](https://docs.google.com/presentation/d/1IrrQaE-jHyzOtFRabjtLAzeP5UOirFOEj8FRADAceqk/edit?usp=sharing). The purpose of this document is to delve deeper into that proposal, offering a comprehensive design of the KHR audio graph. This includes a detailed description of each node object within the graph along with functionality, the specific properties associated with it, and how it interacts with other nodes in the graph. The document is structured to facilitate clear understanding and to solicit feedback on the proposed design. Based on the feedback we will update and finalize the design before it is formally schematized into KHR core audio spec, extensions, animations, and interactivity.
```

   </td>
   <td>object
   </td>
   <td>See 4.4 Encoding properties.
   </td>
   <td>
   </td>
   <td>Could be a part of audio data property  instead.
   </td>
  </tr>
</table>

### 4.2 Audio data

Audio data objects define where audio data is located and what format the data is in. The data is either accessed via a bufferView or uri.

<table>
  <tr>
   <td>
   </td>
   <td><strong>Type</strong>
   </td>
   <td><strong>Description</strong>
   </td>
   <td><strong>Required</strong>
   </td>
   <td><strong>Notes</strong>
   </td>
  </tr>
  <tr>
   <td>buffer view
   </td>
   <td>integer
   </td>
   <td>The index of the buffer view that contains the audio data. The buffer represents an audio asset residing in memory, created from decoding an audio file, or from raw data.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>mime type
   </td>
   <td>string
   </td>
   <td>The audio's MIME type. Required if buffer view is defined.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>uri
   </td>
   <td>string
   </td>
   <td>The uri of the audio file. Relative paths are relative to the .gltf file.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
</table>

> ### 4.3 Oscillator data

Reviewer 2: Has band-limiting been considered for audio rate oscillator implementations? Is it assumed? Or does it need to be explicitly required? I would think it should be explicitly required if audio quality is a priority.

Author: Made that explicit by including a detune parameter. The implementation can then use this with frequency to calculate the desired oscillator frequency.

Reviewer 2: Ok, thanks, but I don't think adding a `detune` parameter has anything to do with making band-limiting explicit. If it's assumed that implementations will apply band-limiting to avoid aliasing, then that's fine, but I suspect many implementations will take the easier route of skipping band-limiting if it's not required, which will reduce audio quality.

It may be that this concern is outside the scope of the standard, but if there a lot of poor-quality implementations then it may cause creators to avoid using it, so we might want to make band-limiting required.

Author: There are various practical methods an implementation might use to prevent aliasing. While the ideal discrete-time digital audio signal is clearly defined mathematically, each implementation must balance the trade-off between computational cost and fidelity to this ideal. It is anticipated that implementations will strive towards this ideal, but it is also reasonable to consider lower-quality, more cost-effective approaches for less powerful hardware. We can include more guidance/recommendation around this.

   </td>
   <td><strong>Notes</strong>
   </td>
  </tr>
  <tr>
   <td>distance model
   </td>
   <td>string
   </td>
   <td>Specifies the distance model for the audio emitter  linear, inverse, exponential, custom).
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>ref distance
   </td>
   <td>number
   </td>
   <td>A reference distance for reducing volume as the emitter moves further from the listener.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>max distance
   </td>
   <td>number
   </td>
   <td>The maximum distance between the emitter and listener, beyond which the audio cannot be heard.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>rolloff factor
   </td>
   <td>number
   </td>
   <td>Describes how quickly the volume is reduced as the emitter moves away from the listener.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
</table>

### 5.4 Shape properties

<table>
  <tr>
   <td>
   </td>
   <td><strong>Type</strong>
   </td>
   <td><strong>Description</strong>
   </td>
   <td><strong>Required</strong>
   </td>
   <td><strong>Notes</strong>
   </td>
  </tr>
  <tr>
   <td>type
   </td>
   <td>string
   </td>

>    <td>Shape in which emitter emits audio (cone, omnidirectional, custom).

Reviewer 2: Why is `custom` included here? Is there a way of defining a custom shape? How would it be implemented?

Reviewer: It would be implemented by another glTF extension, extending this extension.

>    <td>Shape in which emitter emits audio (cone, omnidirectional, custom).

Reviewer 2: Does `omnidirectional` need to be included when it can be achieved with a 360-degree cone angle?

Reviewer: Note: This is how [KHR_audio_emitter](https://github.com/KhronosGroup/glTF/pull/2137) works.

Author: Keeping it explicit to make it runtime friendly.

   </td>
   <td><strong>Description</strong>
   </td>
   <td><strong>Required</strong>
   </td>
   <td><strong>Notes</strong>
   </td>
  </tr>
  <tr>
   <td>input channels
   </td>
   <td>integer
   </td>
   <td>Number of channels in input audio.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>output channels
   </td>
   <td>integer
   </td>
   <td>Number of channels in output audio.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>channel interpretation
   </td>
   <td>string
   </td>
   <td>speakers, discrete
<p>
Speakers use up-mix / down-mix equations. In cases where the number of channels do not match any of the basic speaker layouts, use "discrete". "Discrete" up-mix by filling channels until they run out then zero out remaining channels; down-mix by filling as many channels as possible, then dropping remaining channels.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
</table>

### 6.7 Audio mixer node (N inputs / 1 output)

Use the Audio mixer node to combine the output from multiple audio sources. Number of channels should be the same across all inputs.

[no properties for this node]

### 6.8 Filter node (1 input / 1 output)

> Use the Audio Mixer node to combine the output from multiple audio sources. A filter node always has exactly one input and one output.

Reviewer 2: This looks to be partially copied from the previous item by mistake.

Author: Updated.

   <td>
   </td>
   <td><strong>Type</strong>
   </td>
   <td><strong>Description</strong>
   </td>
   <td><strong>Required</strong>
   </td>
   <td><strong>Notes</strong>
   </td>
  </tr>
  <tr>
   <td>type
   </td>
   <td>string
   </td>
   <td>Defining the kind of filtering algorithm the node is implementing (lowpass, highpass, bandpass, lowshelf, highshelf, peaking, notch, allpass, custom)
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>frequency
   </td>
   <td>number
   </td>
   <td>Frequency in the current filtering algorithm measured in Hz.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>quality factor
   </td>
   <td>number
   </td>
   <td>The lower the Quality, the broader the bandwidth of frequencies cut or boosted. Value range: 0 to 100.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>gain
   </td>
   <td>number
   </td>
   <td>gain applied to the signal by the filter.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>bypass
   </td>
   <td>boolean
   </td>
   <td>Disables this processor while still allowing unprocessed audio signals to pass.
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
</table>

> ### 6.9 Reverb node (1 input / 1 output)

Reviewer 2: Reverb implementations vary widely. Has any thought been given to consistency across implementations for this node? Web-based implementations would have to create their own reverb nodes since there is no WebAudio reverb node, yet. Would a `Convolver` node make more sense?

Author: Good callout, extended support for convolution reverb in addition to algorithmic reverb.

Reviewer 2: Ok, thanks, but this doesn't address my concern about the algorithmic reverb sounding different across implementations. I don't think sound designers are going to use a reverb if it doesn't sound the same everywhere, so how do we get the algorithmic reverb to sound the same on all implementations? Can a specific reverb algorithm be required by the standard, like maybe a basic plate reverb?

Author: Got it. Unfortunately, there is no universal standard for reverb implementation, which can vary significantly in core business logic and its complexity. Many implementations remain unpublished. Therefore, the objective is not to restrict reverb to a particular implementation but to rather parameterize it sufficiently to accommodate most implementations. Requiring a specific reverb algorithm as a standard would, in fact, diminish the value for users. That said, one can always leverage convolution reverb with a specific impulse response.

Reviewer 2: > Requiring a specific reverb algorithm as a standard would, in fact, diminish the value for users.

Maybe, but I think users won't create graphs with algorithmic reverbs because they won't know how they'll sound on all implementations. Instead, they'll bake the exact reverb they want directly into the audio sources. They'll do this with any audio nodes that vary significantly across implementations.

Author: >because they won't know how they'll sound on all implementations 

That's a valid point, and it extends beyond just different implementations to the same implementation being used across various device types, such as mobiles, desktops, and HMDs. Each device type, and indeed each variant within those types, can potentially alter the sound characteristics, making it challenging for users to predict how the reverb will sound without testing it on each device themselves. Moreover, developers often face a trade-off between computational cost and fidelity in their implementations. This balance is crucial to ensure that the reverb not only meets quality expectations but also remains computationally feasible and customizable across diverse hardware platforms.
